python_modules/dagster/dagster/_core/definitions/asset_spec.py

```diff
@@ -0,0 +1,117 @@
+from typing import AbstractSet, Any, Iterable, Mapping, NamedTuple, Optional, Union
+
+import dagster._check as check
+from dagster._annotations import PublicAttr
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.backfill_policy import BackfillPolicy
+from dagster._core.definitions.source_asset import SourceAsset
+
+from .auto_materialize_policy import AutoMaterializePolicy
+from .events import (
+    AssetKey,
+    CoercibleToAssetKey,
+)
+from .freshness_policy import FreshnessPolicy
+from .metadata import MetadataUserInput
+
+
+class AssetSpec(
+    NamedTuple(
+        "_AssetSpec",
+        [
+            ("asset_key", PublicAttr[AssetKey]),
+            ("deps", PublicAttr[AbstractSet[AssetKey]]),
+            ("description", PublicAttr[Optional[str]]),
+            ("metadata", PublicAttr[Optional[Mapping[str, Any]]]),
+            ("group_name", PublicAttr[Optional[str]]),
+            ("skippable", PublicAttr[bool]),
+            ("code_version", PublicAttr[Optional[str]]),
+            ("freshness_policy", PublicAttr[Optional[FreshnessPolicy]]),
+            ("auto_materialize_policy", PublicAttr[Optional[AutoMaterializePolicy]]),
+            ("backfill_policy", PublicAttr[Optional[BackfillPolicy]]),
+        ],
+    )
+):
+    """Specifies the core attributes of an asset. This object is attached to the decorated
+    function that defines how it materialized.
+
+    Attributes:
+        asset_key (AssetKey): The unique identifier for this asset.
+        deps (Optional[AbstractSet[AssetKey]]): The asset keys for the upstream assets that
+            materializing this asset depends on.
+        description (Optional[str]): Human-readable description of this asset.
+        metadata (Optional[Dict[str, Any]]): A dict of static metadata for this asset.
+            For example, users can provide information about the database table this
+            asset corresponds to.
+        skippable (bool): Whether this asset can be omitted during materialization, causing downstream
+            dependencies to skip.
+        group_name (Optional[str]): A string name used to organize multiple assets into groups. If
+            not provided, the name "default" is used.
+        code_version (Optional[str]): The version of the code for this specific asset,
+            overriding the code version of the materialization function
+        freshness_policy (Optional[FreshnessPolicy]): A policy which indicates how up to date this
+            asset is intended to be.
+        auto_materialize_policy (Optional[AutoMaterializePolicy]): AutoMaterializePolicy to apply to
+            the specified asset.
+        backfill_policy (Optional[BackfillPolicy]): BackfillPolicy to apply to the specified asset.
```

Comment:
I think this should go on the op, not on the asset.

Reply:
nice, will rm